### ISSUE TITLE
Allow preProcess, postProcess, afterAll to await promises

### ISSUE THREAD

**Issue by lanthaler** [w3c/respec#1034]
It looks like the current version from https://www.w3.org/Tools/respec/respec-w3c-common broke [our spec](http://www.hydra-cg.com/spec/latest/core/):

```
TypeError: t is not a function
    at https://www.w3.org/Tools/respec/respec-w3c-common:1:27240
    at Array.forEach (native)
    at once (https://www.w3.org/Tools/respec/respec-w3c-common:1:27213)
    at https://www.w3.org/Tools/respec/respec-w3c-common:1:26203
    at https://www.w3.org/Tools/respec/respec-w3c-common:1:25876
    at Array.forEach (native)
    at t (https://www.w3.org/Tools/respec/respec-w3c-common:1:25847)
    at Object.<anonymous> (https://www.w3.org/Tools/respec/respec-w3c-common:2:2699)
    at n (https://www.w3.org/Tools/respec/respec-w3c-common:5:2644)
    at Generator._invoke (https://www.w3.org/Tools/respec/respec-w3c-common:5:4071)
(anonymous) @ linter.js:68
```

**Comment by marcoscaceres** [comment]
@lanthaler I did some investigation... if I disable "preProc" in your ReSpec config, it works... so it means that your `preProc` function/object is doing something strange. Please note that we don't really support or encourage people use `preProc`, as it's supposed to be used as an internal/private function. 

Looking at your "preProc" object, I can see that it's going to lead to race conditions... basically, you are asking ReSpec to synchronously run the function, but then you are doing a network request... but ReSpec is going to do it's processing in between the network request being made and data coming back. That will lead to unexpected results.  Also, preProcess expects functions, not objects. 

I would suggest change your code to (or something similar):

```JS 
// make sure ReSpec loads first
document
  .querySelector('script[src="https://www.w3.org/Tools/respec/respec-w3c-common"]')
  .onload = function(){
    // make sure ReSpec is done.... 
    document.respecIsReady.then(function(){
      if (0 === $('#vocabulary-classes').length) {
        return;
      }
      $.getJSON('core.jsonld', function(vocab) {
        // do processing here. 
      })
    });
  };
```

**Comment by lanthaler** [comment]
Thanks for looking into it @marcoscaceres. Unfortunately I didn't get your proposed code to work. I switched to `postProcess` to avoid the race conditions you described.

Btw. if you don't want `preProc` to be used the documentation at https://github.com/w3c/respec/wiki/preProcess should be updated. To me it looks like a stable config option like any other.

**Comment by marcoscaceres** [comment]
Thanks, and good point about the docs... I'll update them... leaving this open till I do.

**Comment by marcoscaceres** [comment]
@lanthaler, I've given this a bit more thought... I think what I will do is make "preProcess" a bit smarter by allowing it to "await" any promises returned. That will wait for the preProcess list to complete, allowing for async pre-processing of the document before ReSpec's plugins run.  

That will allow us to keep preProcess as a public interface that is backwards compatible, and will work well for others in the future.

So, basically: 

```JS
function doThisAsyncBeforeProcessing(){
   return new Promise(function(resolve){
     doAsyncStuff().then(resolve);
  });
}
``

**Comment by lanthaler** [comment]
Great idea and thanks for implementing it @marcoscaceres.
